spec/rspec/rails/matchers/have_reported_error_spec.rb

```diff
@@ -0,0 +1,233 @@
+    end
+
+    it "fails when error message does not match pattern" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("error without match"))
+      end
+      matcher = have_reported_error(/different pattern/)
+
+      expect(matcher.matches?(test_block)).to be false
+    end
+  end
+
+  describe "symbol error matching" do
+    it "passes when symbol matches" do
+      test_block = proc do
+        Rails.error.report(:test_symbol)
+      end
+
+      expect(test_block).to have_reported_error(:test_symbol)
+    end
+
+    it "fails when symbol does not match" do
+      test_block = proc do
+        Rails.error.report(:actual_symbol)
+      end
+      matcher = have_reported_error(:expected_symbol)
+
+      expect(matcher.matches?(test_block)).to be false
+    end
+  end
+
+  describe "attribute matching with .with chain" do
+    it "passes when attributes match exactly" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"), user_id: 123, context: "test")
+      end
+
+      expect(test_block).to have_reported_error.with(user_id: 123, context: "test")
+    end
+
+    it "passes with partial attribute matching" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"), user_id: 123, context: "test", extra: "data")
+      end
+
+      expect(test_block).to have_reported_error.with(user_id: 123)
+    end
+
+    it "passes with hash matching using RSpec matchers" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"), params: { foo: "bar", baz: "qux" })
+      end
+
+      expect(test_block).to have_reported_error.with(params: a_hash_including(foo: "bar"))
+    end
+
+    it "fails when attributes do not match" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"), user_id: 123, context: "actual")
+      end
+      matcher = have_reported_error.with(user_id: 456, context: "expected")
+
+      expect(matcher.matches?(test_block)).to be false
+    end
+
+    it "fails when no error is reported but attributes are expected" do
+      test_block = proc { "no error" }
+      matcher = have_reported_error.with(user_id: 123)
+
+      expect(matcher.matches?(test_block)).to be false
+    end
+  end
+
+  describe "cleanup behavior" do
+    it "unsubscribes from error reporter on successful completion" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"))
+      end
+
+      expect(mock_error_reporter).to receive(:unsubscribe)
+
+      expect(test_block).to have_reported_error
+    end
+
+    it "unsubscribes from error reporter even when exception is raised" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"))
+        raise "unexpected error"
+      end
+
+      expect(mock_error_reporter).to receive(:unsubscribe)
+
+      expect {
+        have_reported_error.matches?(test_block)
+      }.to raise_error("unexpected error")
+    end
+  end
+
```

Comment:
This might be tested directly, by adding a spec with expect { } with this matcher.